File: Prog_lab12_java/Human.java
package Prog_lab12_java;

import java.util.*;

public class Human implements Cloneable
{
	//Attributes
	private int id;
	private int age;
	private int height;
	private double weight;
	private char gender;
	public Fio fioField;

	//Constructors
	public Human()
	{
		id = 0;
		age = 0;
		height = 40;
		weight = 0;
		gender = 'M';
		fioField = new Fio();
	}

	public Human(int bufId)
	{
		id = 0;
		age = 0;
		height = 40;
		weight = 0;
		gender = 'M';
		fioField = new Fio();

		setId(bufId);
	}

	public Human(int bufId, int bufAge, int bufHeight, double bufWeight, char bufGender, Fio bufFio)
	{
		Human check = new Human();

		if (check.setId(bufId) || check.setAge(bufAge) || check.setHeight(bufHeight) ||
				check.setWeight(bufWeight) || check.setGender(bufGender))
		{
			id = 0;
			age = 0;
			height = 40;
			weight = 0;
			gender = 'M';
			fioField = new Fio();
		}
		else
		{
			setId(bufId);
			setAge(bufAge);
			setHeight(bufHeight);
			setWeight(bufWeight);
			setGender(bufGender);
			fioField = new Fio(bufFio.getSurname(), bufFio.getName(), bufFio.getPatronymic());
		}
	}

	//Methods with Exceptions
	public void hardSetId(int buf)
	{
		if (buf < 0 || buf > 9999)
			throw new NumberFormatException("Invalid range of number.\n");
		else
			id = buf;
	}

	public void hardSetAge(int buf)
	{
		if (buf < 0 || buf > 200)
			throw new NumberFormatException("Invalid range of number.\n");
		else
			age = buf;
	}

	public void hardSetHeight(int buf)
	{
		if (buf < 40 || buf > 300)
			throw new NumberFormatException("Invalid range of number.\n");
		else
			height = buf;
	}

	public void hardSetWeight(double buf)
	{
		//round *.* format to *.00 format
		buf = (double)(Math.round(buf * 100)) / 100;
		if (buf < 0 || buf > 650)
			throw new NumberFormatException("Invalid range of number.\n");
		else
			weight = buf;
	}

	public void hardSetGender(char buf)
	{
		if (buf == 'M' || buf == 'F')
			gender = buf;
		else
			throw new IllegalStateException("Invalid symbol.\n");
	}


	//Methods
	public int getBodyMassIndex()
	{
		return (BodyMassIndex.calculate(height, weight));
	}

	public boolean setId(int buf)
	{
		if (buf < 0 || buf > 9999)
			return (true);
		else
		{
			id = buf;
			return (false);
		}
	}

	public boolean setAge(int buf)
	{
		if (buf < 0 || buf > 200)
			return (true);
		else
		{
			age = buf;
			return (false);
		}
	}

	public boolean setHeight(int buf)
	{
		if (buf < 40 || buf > 300)
			return (true);
		else
		{
			height = buf;
			return (false);
		}
	}

	public boolean setWeight(double buf)
	{
		//round *.* format to *.1 format 
		buf = (double)(Math.round(buf * 10)) / 10;
		if (buf < 0 || buf > 650)
			return (true);
		else
		{
			weight = buf;
			return (false);
		}
	}

	public boolean setGender(char buf)
	{
		if (buf == 'M' || buf == 'F')
		{
			gender = buf;
			return(false);
		}
		else
			return (true);
	}

	public boolean setAll(int bufId, int bufAge, int bufHeight, double bufWeight, char bufGender, Fio bufFio)
	{
		Human check = new Human();

		if (check.setId(bufId) || check.setAge(bufAge) || check.setHeight(bufHeight) ||
				check.setWeight(bufWeight) || check.setGender(bufGender))
			return (true);
		else
		{
			setId(check.getId());
			setAge(check.getAge());
			setHeight(check.getHeight());
			setWeight(check.getWeight());
			setGender(check.getGender());
			fioField.setAll(bufFio.getSurname(), bufFio.getName(), bufFio.getPatronymic());
			return (false);
		}
	}

	public int getId()
	{
		return (id);
	}

	public int getAge()
	{
		return (age);
	}

	public int getHeight()
	{
		return (height);
	}

	public double getWeight()
	{
		return (weight);
	}

	public char getGender()
	{
		return (gender);
	}

	public void read()
	{
		Human check = new Human();
		Scanner scanner = new Scanner(System.in);

		boolean readFlag = true;
		while(readFlag)
		{
			System.out.print("Enter id:\n");
			try
			{
				check.hardSetId(Integer.parseInt(scanner.nextLine()));
				readFlag = false;
			}
			catch (Exception ex)
			{
				System.out.print(ex);
			}
		}

		readFlag = true;
		while(readFlag)
		{
			System.out.print("Enter age:\n");
			try
			{
				check.hardSetAge(Integer.parseInt(scanner.nextLine()));
				readFlag = false;
			}
			catch (Exception ex)
			{
				System.out.print(ex);
			}
		}

		readFlag = true;
		while(readFlag)
		{
			System.out.print("Enter height:\n");
			try
			{
				check.hardSetHeight(Integer.parseInt(scanner.nextLine()));
				readFlag = false;
			}
			catch (Exception ex)
			{
				System.out.print(ex);
			}
		}

		readFlag = true;
		while(readFlag)
		{
			System.out.print("Enter weight:\n");
			try
			{
				check.hardSetWeight(Double.parseDouble(scanner.nextLine()));
				readFlag = false;
			}
			catch (Exception ex)
			{
				System.out.print(ex);
			}
		}

		readFlag = true;
		while(readFlag)
		{
			System.out.print("Enter gender:\n");
			try
			{
				check.hardSetGender(scanner.next(".").charAt(0));
				readFlag = false;
			}
			catch (Exception ex)
			{
				System.out.print(ex);
			}
		}

		check.fioField.read();

		this.setAll(check.getId(), check.getAge(), check.getHeight(),
				check.getWeight(), check.getGender(), check.fioField);
	}

	public String toString()
	{
	String returnValue = "";
	returnValue += "id: " + id +
			"\nage: " + age +
			"\nheight: " + height +
			"\nweigth: " + weight +
			"\ngender: " + gender +
			"\n" + fioField.toString();
	return (returnValue);
	}

	public Object clone()
	{
		try
		{
			Human clone = (Human)super.clone();
			clone.fioField = (Fio)fioField.clone();
			return (clone);
		}
		catch(CloneNotSupportedException e)
		{

		}
		return this;
	}
}

File: Prog_lab12_csharp/Human.cs
using System;

namespace Prog_lab12
{
    public class Human : ICloneable
    {
		//Attributes
		private int id;
		private int age;
		private int height;
		private double weight;
		private char gender;
		public Fio fioField;

		//Properties
		public int Id
		{
			get
			{
				return id;
			}
			set
			{
				if ((value >= 0) && (value <= 9999))
					id = value;
			}
		}
		public int Age 
		{
			get
			{
				return age;
			}
			set
			{
				if ((value >= 0) && (value <= 200))
					age = value;
			}
		}
		public int Height
		{
			get
			{
				return height;
			}
			set
			{
				if ((value >= 40) && (value <= 300))
					height = value;
			}
		}
		public double Weight
		{
			get
			{
				return weight;
			}
			set
			{
				value = Math.Round(value, 1, MidpointRounding.AwayFromZero);
				if ((value >= 0) && (value <= 650))
					weight = value;
			}
		}
		public char Gender
        {
			get
			{
				return gender;
			}
			set
			{
				if (value == 'M' || value == 'F')
					gender = value;
			}
		}
		public Fio FioField 
		{
			get
			{
				return (fioField);
			}
			set
			{
				fioField = new Fio(value.GetSurname(), value.GetName(), value.GetPatronymic());
			}
		}

		//Constructors
		public Human()
		{
			id = 0;
			age = 0;
			height = 40;
			weight = 0;
			gender = 'M';
			fioField = new Fio();
		}
		public Human(int bufId)
		{
			id = 0;
			age = 0;
			height = 40;
			weight = 0;
			gender = 'M';
			fioField = new Fio();

			SetId(bufId);
		}
		public Human(int bufId, int bufAge, int bufHeight, double bufWeight, char bufGender, Fio bufFio)
		{
			Human check = new Human();

			if (check.SetId(bufId) || check.SetAge(bufAge) || check.SetHeight(bufHeight) ||
				check.SetWeight(bufWeight) || check.SetGender(bufGender))
			{
				id = 0;
				age = 0;
				height = 40;
				weight = 0;
				gender = 'M';
				fioField = new Fio();
			}
			else
			{
				SetId(bufId);
				SetAge(bufAge);
				SetHeight(bufHeight);
				SetWeight(bufWeight);
				SetGender(bufGender);
				fioField = new Fio(bufFio.GetSurname(), bufFio.GetName(), bufFio.GetPatronymic());
			}
		}

		//Methods with Exceptions
		public void HardSetId(int buf)
		{
			if (buf < 0 || buf > 9999)
				throw new Exception("Invalid range of number.");
			else
				id = buf;
		}
		public void HardSetAge(int buf)
		{
			if (buf < 0 || buf > 200)
				throw new Exception("Invalid range of number.");
			else
				age = buf;
		}
		public void HardSetHeight(int buf)
		{
			if (buf < 40 || buf > 300)
				throw new Exception("Invalid range of number.");
			else
				height = buf;
		}
		public void HardSetWeight(double buf)
		{
			//round *.* format to *.00 format 
			buf = Math.Round(buf, 2, MidpointRounding.AwayFromZero);
			if (buf < 0 || buf > 650)
				throw new Exception("Invalid range of number.");
			else
				weight = buf;
		}
		public void HardSetGender(char buf)
		{
			if (buf == 'M' || buf == 'F')
			{
				gender = buf;
			}
			else
				throw new Exception("Invalid symbol.");
		}

		//Methods
		public bool SetId(int buf)
		{
			if (buf < 0 || buf > 9999)
				return (true);
			else
			{
				id = buf;
				return (false);
			}
		}
		public bool SetAge(int buf)
		{
			if (buf < 0 || buf > 200)
				return (true);
			else
			{
				age = buf;
				return (false);
			}
		}
		public bool SetHeight(int buf)
		{
			if (buf < 40 || buf > 300)
				return (true);
			else
			{
				height = buf;
				return (false);
			}
		}
		public bool SetWeight(double buf)
		{
			//round *.* format to *.1 format 
			buf = Math.Round(buf, 1, MidpointRounding.AwayFromZero);
			if (buf < 0 || buf > 650)
				return (true);
			else
			{
				weight = buf;
				return (false);
			}
		}
		public bool SetGender(char buf)
		{
			if (buf == 'M' || buf == 'F')
			{
				gender = buf;
				return (false);
			}
			else
				return (true);
		}
		public bool SetAll(int bufId, int bufAge, int bufHeight, double bufWeight, char bufGender, Fio bufFio)
		{
			Human check = new Human();

			if (check.SetId(bufId) || check.SetAge(bufAge) || check.SetHeight(bufHeight) ||
				check.SetWeight(bufWeight) || check.SetGender(bufGender))
				return (true);
			else
			{
				SetId(check.GetId());
				SetAge(check.GetAge());
				SetHeight(check.GetHeight());
				SetWeight(check.GetWeight());
				SetGender(check.GetGender());
				fioField.SetAll(bufFio.GetSurname(), bufFio.GetName(), bufFio.GetPatronymic());
				return (false);
			}
		}

		public int GetId()
		{ return (id); }
		public int GetAge()
		{ return (age); }
		public int GetHeight()
		{ return (height); }
		public double GetWeight()
		{ return (weight); }
		public char GetGender()
		{ return (gender); }

		public void Read()
		{
			Human check = new Human();
			int bufInt;
			double bufDouble;
			char bufChar;

			bool readFlag = true;
			while (readFlag)
			{
				Console.Write("Enter id:\n");
                try
                {
					bufInt = int.Parse(Console.ReadLine());
					check.HardSetId(bufInt);
					readFlag = false;
				}
                catch (Exception ex)
                {
					Console.WriteLine(ex.Message);
				}
			}

			readFlag = true;
			while (readFlag)
			{
				Console.Write("Enter age:\n");
				try
				{
					bufInt = int.Parse(Console.ReadLine());
					check.HardSetAge(bufInt);
					readFlag = false;
				}
				catch (Exception ex)
				{
					Console.WriteLine(ex.Message);
				}
			}

			readFlag = true;
			while (readFlag)
			{
				Console.Write("Enter height:\n");
				try
				{
					bufInt = int.Parse(Console.ReadLine());
					check.HardSetHeight(bufInt);
					readFlag = false;
				}
				catch (Exception ex)
				{
					Console.WriteLine(ex.Message);
				}
			}

			readFlag = true;
			while (readFlag)
			{
				Console.Write("Enter weight:\n");
				try
				{
					bufDouble = double.Parse(Console.ReadLine());
					check.HardSetWeight(bufDouble);
					readFlag = false;
				}
				catch (Exception ex)
				{
					Console.WriteLine(ex.Message);
				}
			}

			readFlag = true;
			while (readFlag)
			{
				Console.Write("Enter gender:\n");
				try
				{
					bufChar = char.Parse(Console.ReadLine());
					check.HardSetGender(bufChar);
					readFlag = false;
				}
				catch (Exception ex)
				{
					Console.WriteLine(ex.Message);
				}
			}

			check.fioField.Read();

			SetAll(check.GetId(), check.GetAge(), 
				check.GetHeight(), check.GetWeight(), 
				check.GetGender(), check.fioField);
		}
        public override string ToString()
        {
			string returnValue = "";
			returnValue += "id: " + id.ToString() + 
				"\nage: " + age.ToString() + 
				"\nheight: " + height.ToString() + 
				"\nweigth: " + weight.ToString() + 
				"\ngender: " + gender + 
				"\n" + fioField.ToString();
			return (returnValue);
        }
		public object Clone()
		{
			return new Human(id, age, height, weight, gender, fioField);
		}
	}
}
